README.md
# balenaOS device proxy tunnel
> [redirect](https://www.balena.io/docs/reference/OS/network/2.x/#connecting-behind-a-proxy) all TCP traffic from balenaOS devices via an SSH tunnel to a SOCKS5 proxy, running on DigitalOcean or AWS/EC2
> 

[![](https://balena.io/deploy.svg)](https://dashboard.balena-cloud.com/deploy?repoUrl=https://github.com/balena-labs-projects/proxy-tunnel)


## ToC
* [overview](#overview)
* [generate keys](#generate-keys)
* [cloud proxy server](#cloud-proxy-server)
* [balenaCloud application](#balenacloud-application)
* [test and verify](#test-and-verify)
* [troubleshooting](#troubleshooting)


## overview
The examples give below demonstrate how to forward/redirect all balenaOS traffic via a 
user provided proxy. In the specific instance, we use a SOCKS5 proxy, which we tunnel to 
over SSH on DigitalOcean and AWS/EC2 using public key authentication.

Configuration of the hostOS is done within a ~10MB container on the host network. This 
container interacts with the [Supervisor API](https://www.balena.io/docs/reference/supervisor/supervisor-api/) 
to configure the redsocks redirector. As well as configuring the redirector, the 
container optionally establishes a tunnel to a remote server via SSH, if a private key is 
provided.

The redirection on the host is performed by [redsocks](https://github.com/darkk/redsocks) 
using iptables. Redsocks supports `socks4`, `socks5`, `http-connect`, and `http-relay` proxy 
types, and while in these examples we use `socks5` (default), any supported proxy type 
can be deployed and used.

In our example, we do not enable proxy authentication, since we are already tunneling to 
it via SSH, however in case where such local tunneling may not be desirable, setting 
`PROXY_LOGIN` and `PROXY_PASSWORD` environment variables and omitting `SSH_PRIVATE_KEY` 
will not create a tunnel and assume direct communication with the proxy using provided 
proxy authentication credentials.

The general approach uses in this example is as follows:

* decide which cloud provider to use for hosting the proxy server (i.e. Digital Ocean or AWS)
* generate test SSH keys, using an appropriate format for the selected cloud provider
* deploy a proxy server using Terraform into the selected cloud provider
* create a balenaCloud application
* build and deploy a release
* provision a balenaOS device (e.g balenaFin, RaspberryPi, Intel NUC, etc.)
* test and verify
* destroy test resources to avoid unexpected billing


## generate keys
> generate SSH keys for proxy tunnel authentication

    mkdir -p keys


### EC
> EC keys can be used with DigitalOcean servers

    ssh_key_type=ed25519

    ssh-keygen -o -a 100 -t "${ssh_key_type}" -f "keys/id_${ssh_key_type}" -C 'balena' -N ''

    ssh_public_key="$(cat keys/id_ed25519.pub)"

    ssh_private_key="$(cat keys/id_ed25519 | openssl base64 -A)"


### RSA
> used for AWS/EC2 since EC cryptography is not supported at the time of writing

    ssh_key_type=rsa

    ssh-keygen -t "${ssh_key_type}" -f "keys/id_${ssh_key_type}" -C 'balena' -N ''

    ssh_public_key="$(cat keys/id_rsa.pub)"

    ssh_private_key="$(cat keys/id_rsa | openssl base64 -A)"


## cloud proxy server
> deploy a proxy server into a desired cloud provider

### DigitalOcean
> deploy a [DigitalOcean](https://cloud.digitalocean.com/account/api/tokens) cloud proxy server using [Terraform](https://www.terraform.io/downloads.html)

### create proxy server

    export DO_TOKEN={{ digital-ocean-api-token }}
    
    pushd terraform/digitalocean
    
    terraform init

    terraform plan -var "do_token=${DO_TOKEN}"  
  
    terraform apply -var "do_token=${DO_TOKEN}"
    
    proxy_ip="$(terraform output -json | jq -r .ipv4_address.value)"

    ssh_user=root

    popd

#### (finally) clean up resources
> optionally destroy DigitalOcean Droplet after finishing with the PoC

    pushd terraform/digitalocean

    terraform destroy -var "do_token=${DO_TOKEN}"

    popd


### AWS/EC2
> deploy a cloud proxy server into AWS/EC2 using Terraform

#### create proxy server
> ensure your AWS credentials are configured correctly

    pushd terraform/aws-ec2

    terraform init

    terraform plan -var "key_pair=${ssh_public_key}"

    terraform apply -var "key_pair=${ssh_public_key}"

    proxy_ip="$(terraform output -json | jq -r .instance_public_ip.value)"
    
    ssh_user=ec2-user

    popd


#### (finally) clean up resources
> optionally destroy AWS/EC2 resources after finishing with the PoC

    pushd terraform/aws-ec2

    terraform destroy -var "key_pair=${ssh_public_key}"

    popd


## balenaCloud application
> deploy balenaCloud application, which will configure balenaOS to proxy all traffic via your cloud proxy

### build and deploy the app
> once the app is created and the release is deployed, [provision](https://www.balena.io/docs/learn/getting-started/raspberrypi3/nodejs/#add-your-first-device) at least one balenaOS device into it

    balena login

    # to build to other architectures, also change ALPINE_ARCH in docker-compose.yml
    arch=armv7l

    device_type=fincm3

    app_name=${device_type}-${arch}

    balena app create ${app_name} --type ${device_type}

    app_slug=$(balena app ${app_name} | grep SLUG | awk '{print $2}')

    balena env add SSH_USER "${ssh_user}" --application ${app_slug}

    balena env add SSH_KEY_TYPE "${ssh_key_type}" --application ${app_slug}

    balena env add SSH_PRIVATE_KEY "${ssh_private_key}" --application ${app_slug}

    balena env add PROXY_IP "${proxy_ip}" --application ${app_slug}

    balena push ${app_slug}


## test and verify
> terminal into the hostOS container and verify all traffic is exiting via your proxy

### DigitalOcean

    echo 'curl -s https://ipinfo.io/org; exit' \
      | balena ssh $(balena devices --app ${app_name} -j | jq -r .[0].uuid)

	=============================================================
		Welcome to balenaOS
	=============================================================
	AS14061 DigitalOcean, LLC


### AWS/EC2

    ...
    
	=============================================================
		Welcome to balenaOS
	=============================================================
	AS16509 Amazon.com, Inc.


## troubleshooting

* test proxy connectivity bypassing redsocks redirector on the balenaOS device

```
curl -x socks5://127.0.0.1:1080 https://ipinfo.io/
```

* to temporarily restore access in the event of proxy connectivity issues, on the hostOS run the following

```
iptables -t nat -F OUTPUT
iptables -t nat -F PREROUTING
iptables -t nat -F REDSOCKS
```

---
Icon made by [Freepik](https://www.freepik.com "Freepik") from [www.flaticon.com](https://www.flaticon.com/ "Flaticon").


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v0.9.3
## (2023-09-27)

* Remove note for excluding VPN traffic [Zahari Petkov]

# v0.9.2
## (2023-08-23)

* Update Terraform terraform-aws-modules/security-group/aws to v4.17.2 [Self-hosted Renovate Bot]

# v0.9.1
## (2023-08-23)

* Update Terraform digitalocean to v2.29.0 [Self-hosted Renovate Bot]

# v0.9.0
## (2023-08-23)

* Update Terraform terraform-aws-modules/ec2-instance/aws to v4.5.0 [Self-hosted Renovate Bot]

# v0.8.2
## (2022-10-24)

* Updating repo reference from balenalabs to balena-labs-projects [Flynn Joffray]

# v0.8.1
## (2022-10-10)

* Switch from balenaCI to flowzone [Pagan Gazzard]

# v0.8.0
## (2022-10-10)

* Update Terraform digitalocean to 2.23.0 [Renovate Bot]

# v0.7.3
## (2022-09-26)

* Update Terraform terraform-aws-modules/security-group/aws to 4.13.1 [Renovate Bot]

# v0.7.2
## (2022-09-12)

* Update Terraform digitalocean to 2.22.3 [Renovate Bot]

# v0.7.1
## (2022-08-31)

* Update Terraform digitalocean to 2.22.2 [Renovate Bot]

# v0.7.0
## (2022-08-23)

* Update Terraform terraform-aws-modules/security-group/aws to 4.13.0 [Renovate Bot]

# v0.6.0
## (2022-08-22)

* Update Terraform terraform-aws-modules/security-group/aws to 4.12.0 [Renovate Bot]

# v0.5.0
## (2022-08-18)

* Update Terraform terraform-aws-modules/security-group/aws to 4.11.0 [Renovate Bot]

# v0.4.0
## (2022-08-16)

* Update Terraform digitalocean to 2.22.1 [Renovate Bot]

# v0.3.0
## (2022-08-16)

* Update Terraform terraform-aws-modules/security-group/aws to 4.10.0 [Renovate Bot]

# v0.2.4
## (2022-08-13)

* Update Terraform terraform-aws-modules/ec2-instance/aws to 4.1.4 [Renovate Bot]

# v0.2.3
## (2022-08-12)

* Update Terraform terraform-aws-modules/ec2-instance/aws to 4.1.3 [Renovate Bot]

# v0.2.2
## (2022-08-10)

* Update Terraform terraform-aws-modules/ec2-instance/aws to 4.1.2 [Renovate Bot]

# v0.2.1
## (2022-07-21)

* Update Terraform terraform-aws-modules/ec2-instance/aws to 4.1.1 [Renovate Bot]

# v0.2.0
## (2022-07-19)

* Update Terraform terraform-aws-modules/ec2-instance/aws to 4.1.0 [Renovate Bot]

# v0.1.0
## (2022-06-30)

* Update Terraform digitalocean from 2.20.0 to 2.21.0 [Renovate Bot]

# v0.0.14
## (2022-06-14)

* Update Terraform digitalocean from 1.23.0 to 2.20.0 [Renovate Bot]

# v0.0.13
## (2022-06-14)

* Update Terraform terraform-aws-modules/ec2-instance/aws from 2.21.0 to 4.0.0 [Renovate Bot]

# v0.0.12
## (2022-06-13)

* Update Terraform terraform-aws-modules/security-group/aws from 3.18.0 to 4.9.0 [Renovate Bot]
* Update Terraform aws from 3.75.2 to 4.18.0 [Renovate Bot]

# v0.0.11
## (2022-06-13)

* Update Terraform hashicorp/terraform from 0.13.7 to 1.2.2 [Renovate Bot]

# v0.0.10
## (2022-06-13)

* Update Terraform terraform-aws-modules/ec2-instance/aws from 2.17.0 to 2.21.0 [Renovate Bot]
* Update Terraform digitalocean from 1.22.2 to 1.23.0 [Renovate Bot]

# v0.0.9
## (2021-05-13)

* DWB button added on the README [Marc Pous]

# v0.0.8
## (2021-05-06)

* Reword contract and readme [ab77]

# v0.0.7
## (2021-05-06)

* Include balena contract in release [ab77]

# v0.0.6
## (2021-05-06)

* Delete superfluous template [ab77]

# v0.0.5
## (2021-05-06)

* Rename Dockerfile [ab77]

# v0.0.4
## (2021-05-06)

* Test balenaCI pipeline [ab77]

# v0.0.3
## (2021-05-05)

* Change to balena repo type [ab77]

# v0.0.2
## (2021-05-05)

* Update documentation [ab77]
* Remove balenaCI docker pipeline matadata [ab77]
* Refactor for balena push [ab77]
* Add note about VPN bypass [ab77]
* Change to use balena push workflow [ab77]
* Set versions [ab77]
* Ignore lock file(s) [ab77]
* Disable redirection in proxy_ip is empty [ab77]
* Fix instructions [ab77]
* Add note [ab77]
* Markdown formatting [ab77]
* Initial commit [ab77]


app/balena.sh
#!/bin/sh

set -eu

# https://www.balena.io/docs/reference/OS/network/2.x/#connecting-behind-a-proxy
proxy_ipv4=${PROXY_IP:-}
proxy_user=${PROXY_LOGIN:-}
proxy_pass=${PROXY_PASSWORD:-}
proxy_port=${PROXY_PORT:-1080}
proxy_type=${PROXY_TYPE:-socks5}
ssh_user=${SSH_USER:-root}
ssh_port=${SSH_PORT:-22}
ssh_private_key=${SSH_PRIVATE_KEY:-}
ssh_key_type=${SSH_KEY_TYPE:-}


# if provided, decode and store SSH key material
if [ "${ssh_private_key}" != '' ] && [ "${ssh_key_type}" != '' ]; then
    mkdir -p "${HOME}/.ssh"
    ssh_private_key_file="${HOME}/.ssh/id_${ssh_key_type}"
    touch "${ssh_private_key_file}"
    chmod 0600 "${ssh_private_key_file}"
    echo "${ssh_private_key}" | base64 -d > "${ssh_private_key_file}"
fi

# https://www.balena.io/docs/reference/supervisor/supervisor-api/#patch-v1devicehost-config
if [ "${proxy_ipv4}" != '' ]; then
    tmpconfig="$(mktemp)"
    config="$(mktemp)"
    echo "{\"network\":{\"proxy\":{\"noProxy\":[\"${proxy_ipv4}\"],\"type\":\"${proxy_type}\",\"ip\":\"${proxy_ipv4}\",\"port\":${proxy_port}}}}" | jq -r > "${config}"
    cat < "${config}" > "${tmpconfig}"
    
    # handle proxy authentication
    if [ "${proxy_user}" != '' ] && [ "${proxy_pass}" != '' ]; then
        cat < "${config}" | jq ".network.proxy += {\"login\":\"$proxy_user\"}" > "${tmpconfig}"
        cat < "${tmpconfig}" > "${config}"
        cat < "${config}" | jq ".network.proxy += {\"password\":\"$proxy_pass\"}" > "${tmpconfig}"
        cat < "${tmpconfig}" > "${config}"
    fi
    
    # handle SSH tunnel
    [ -f "${ssh_private_key_file}" ] &&  cat < "${config}" \
      | jq ".network.proxy += {\"ip\":\"127.0.0.1\"}" > "${tmpconfig}"
    cat < "${tmpconfig}" > "${config}"

	# enable redsocks redirector
    cat < "${config}" | jq && curl -X PATCH "${BALENA_SUPERVISOR_ADDRESS}/v1/device/host-config?apikey=${BALENA_SUPERVISOR_API_KEY}" \
      --header 'Content-Type:application/json' \
      --data "$(cat "${config}")"
else
	# disable redsocks redirector
    curl -X PATCH "${BALENA_SUPERVISOR_ADDRESS}/v1/device/host-config?apikey=${BALENA_SUPERVISOR_API_KEY}" \
      --header 'Content-Type:application/json' \
      --data '{}'
fi

# open SSH tunnel to a SOCKS proxy if all the conditions are met
if [ "${proxy_ipv4}" != '' ] && [ -f "${ssh_private_key_file}" ]; then
  exec ssh -vCND "127.0.0.1:${proxy_port}" \
    -o StrictHostKeyChecking=no \
    -i "${ssh_private_key_file}" \
    -p "${ssh_port}" \
    "${ssh_user}@${proxy_ipv4}" "@$"
fi
